sdks/go/examples/wasm/wasm.go

```diff
@@ -0,0 +1,159 @@
+		return fmt.Errorf("--output is required")
+	}
+	return nil
+}
+
+func main() {
+	flag.Parse()
+	beam.Init()
+	ctx := context.Background()
+	if err := preRun(); err != nil {
+		panic(err)
+	}
+	if err := run(ctx); err != nil {
+		panic(err)
+	}
+}
+
+func run(ctx context.Context) error {
+	p, s := beam.NewPipelineWithRoot()
+
+	in := beam.Create(s, "a", "b", "c", "d", "e", "f", "g", "h")
+
+	out := beam.ParDo(s.Scope("embeddedWasm"), &embeddedWasmFn{}, in)
+
+	textio.Write(s, *output, out)
+
+	if err := beamx.Run(ctx, p); err != nil {
+		return fmt.Errorf("failed to run pipeline: %v", err)
+	}
+	return nil
+}
+
+// Concept #2 wrap wasm function execution within a DoFn.
+// wasmFn wraps a DoFn to execute a Rust compiled wasm function
+type embeddedWasmFn struct {
+	r                              wazero.Runtime
+	mod                            api.Module
+	greeting, allocate, deallocate api.Function
+}
+
+// Setup loads and initializes the embedded wasm functions
+// Concept #1: Load a compiled wasm file []byte content and function.
+// This example is derived from
+// https://github.com/tetratelabs/wazero/blob/v1.0.0-pre.3/examples/allocation/rust/greet.go
+func (fn *embeddedWasmFn) Setup(ctx context.Context) error {
+	fn.r = wazero.NewRuntime(ctx)
+	_, err := fn.r.NewHostModuleBuilder("env").
+		NewFunctionBuilder().WithFunc(logString).Export("log").
+		Instantiate(ctx, fn.r)
+	if err != nil {
+		return fmt.Errorf("failed to instantiate host module: %w", err)
+	}
+	fn.mod, err = fn.r.InstantiateModuleFromBinary(ctx, greetWasm)
+	if err != nil {
+		return fmt.Errorf("failed to instantiate wasm module: %v", err)
+	}
+	fn.greeting = fn.mod.ExportedFunction(wasmFunctionName)
+	fn.allocate = fn.mod.ExportedFunction(wasmAllocateFunctionName)
+	fn.deallocate = fn.mod.ExportedFunction(wasmDeallocateFunctionName)
+	return nil
+}
+
+// ProcessElement processes a string calling a wasm function written in Rust
+// This example is derived from
+// https://github.com/tetratelabs/wazero/blob/v1.0.0-pre.3/examples/allocation/rust/greet.go
+func (fn *embeddedWasmFn) ProcessElement(ctx context.Context, s string) (string, error) {
+	size := uint64(len(s))
+	results, err := fn.allocate.Call(ctx, size)
+	if err != nil {
+		return "", fmt.Errorf("error calling allocate: %w", err)
+	}
+	ptr := results[0]
+	defer fn.deallocate.Call(ctx, ptr, size)
+	if !fn.mod.Memory().Write(ctx, uint32(ptr), []byte(s)) {
+		return "", fmt.Errorf("Memory.Write(%d, %d) out of range of memory size %d",
+			ptr, size, fn.mod.Memory().Size(ctx))
+	}
+
+	ptrSize, err := fn.greeting.Call(ctx, ptr, size)
+	resultPtr := uint32(ptrSize[0] >> 32)
+	resultSize := uint32(ptrSize[0])
+	defer fn.deallocate.Call(ctx, uint64(resultPtr), uint64(resultSize))
+	bytes, ok := fn.mod.Memory().Read(ctx, resultPtr, resultSize)
+	if !ok {
+		return "", fmt.Errorf("Memory.Read(%d, %d) out of range of memory size %d",
+			resultPtr, resultSize, fn.mod.Memory().Size(ctx))
+	}
+	return string(bytes), nil
+}
```

Comment:
Given the direct pointer/offset + size uses, it would be good to add comments. I tried just wraping in methods, but that didn't help very much, so I propose doing something like...

```suggestion
func (fn *embeddedWasmFn) ProcessElement(ctx context.Context, s string) (string, error) {
	// Write the string to wasm
	size := uint64(len(s))
	results, err := fn.allocate.Call(ctx, size)
	if err != nil {
		return "", fmt.Errorf("error calling allocate: %w", err)
	}
	ptr := results[0]
	defer fn.deallocate.Call(ctx, ptr, size)
	if !fn.mod.Memory().Write(ctx, uint32(ptr), []byte(s)) {
		return "", fmt.Errorf("Memory.Write(%d, %d) out of range of memory size %d",
			ptr, size, fn.mod.Memory().Size(ctx))
	}

	// Call the wasm function and read the updated string
	ptrSize, err := fn.greeting.Call(ctx, ptr, size)
	resultPtr := uint32(ptrSize[0] >> 32)
	resultSize := uint32(ptrSize[0])
	defer fn.deallocate.Call(ctx, uint64(resultPtr), uint64(resultSize))
	bytes, ok := fn.mod.Memory().Read(ctx, resultPtr, resultSize)
	if !ok {
		return "", fmt.Errorf("Memory.Read(%d, %d) out of range of memory size %d",
			resultPtr, resultSize, fn.mod.Memory().Size(ctx))
	}
	return string(bytes), nil
}
```

Reply:
I added back the original helpful comments from the wazero example with some edits to reflect the Beam context within which the DoFn invokes these methods.

Reply:
Agreed there's a fine line in what to strip out when the crux of the example is "putting chocolate into my peanut butter". 
In this case, wasm is just complicated, so it should be commented appropriately in any of it's contexts, if only because it makes it easier for someone else to get started in the *other* direction. (eg. "I know beam, but would like to learn how to put wasm stuff into my DoFns"). Afterall, by the time you're considering using wasm with beam, you're already familiar enough with Beam.

Anything more complicated than this example is all up to the WASM/Wasmer/Wazero communities ;).